### PR TITLE
Replaced bundled 'pypng' (png) dependency with pillow

### PR DESCRIPTION
**Pull request by Kartik-byte**
With respect to issue #1925 ,this PR replaces pypng with pillow

### CODE REVIEW

File: vispy/io/image.py [hunk 1]
```diff
@@ -117,21 +116,26 @@ def read_png(filename):
...
-        else:
-            y = x.asRGB8()[2]
-            n = 3
-        y = np.array([yy for yy in y], np.uint8)
-    finally:
-        x.file.close()
-    y.shape = (y.shape[0], y.shape[1] // n, n)
-    return y
-
+    imageio, PIL = _check_img_lib()
```
Comment (djhoese) [inline vispy/io/image.py:119]: What do you think about a try/except ImportError with `import PIL` and changing the error message to say Pillow?

File: setup.py [hunk 2]
```diff
@@ -222,7 +222,7 @@ def run(self):
         'sdl2': ['PySDL2'],
         'wx': ['wxPython'],
         'doc': ['sphinx_bootstrap_theme', 'numpydoc'],
-        'io': ['meshio'],
+        'io': ['meshio','PIL'],
```
[the remark this answers is not in the document]
Comment (Kartik-byte) [reply setup.py:225]: Sure, will change that.
Thank you.

File: vispy/io/image.py (hunk 1 above)
Comment (Kartik-byte) [reply vispy/io/image.py]: Okay, will do.

**Comment by Kartik-byte** [comment]
Hey I made the required changes.
Please let me know if something else needs to be changed.
Thank You.

**Review by djhoese — COMMENTED** [review]
Hm the tests are still failing. See if adding `pillow` to every one of the CONDA_DEPENDENCIES in .travis.yml will fix the failing travis tests. Appveyor and azure may need similar treatments. This is a difficult removal of a dependency because although it is used in one place, that function is used all over the tests. 

https://github.com/vispy/vispy/blob/ba54e0a65dcc25352f0f138d92c15140c6da0273/.travis.yml#L17

**Comment by Kartik-byte** [comment]
Hey, Thanks for the quick response.
Sorry I am very new to this, I tried fixing my PR for the tests but still some tests are not still not passing.
I am unable to understand the error messages.
Please guide me on what to do next.
Thank You.

**Comment by djhoese** [comment]
No problem. The tests seem to be complaining about SSL certificates when downloading some test data. This sounds wrong to me and is definitely not something vispy (or you) should have to worry about. I've restarted the failing tests to see if they fail again. The weirdest part is that only the OSX environments failed. :shrug:

**Comment by Kartik-byte** [comment]
Thank you once again, I had no idea what was going on.
Can you let me know whether restarting the tests worked ??
It still shows not passed here ( sorry I am a complete newbie)
If not, is there something I can do (maybe in the azure config file) that can fix this?
I think it might be one of those new extra secure security measures that macs are know for. 
Anyways Thank You.😄

**Comment by djhoese** [comment]
So the error didn't go away after restart, but there is another pull request (running the same tests) that is working fine. I'm not sure how that makes sense. I'll have to look at this later. Kind of short on time right now and I'm not sure where to start with figuring this one out.

**Comment by Kartik-byte** [comment]
No Problem,
Thanks again,
It has to be with something I changed.
No worries I will try to find what is creating this.
I will let you know if it is fixed.
Thanks for the help.😁

File: vispy/util/fetching.py [hunk 3]
```diff
@@ -234,9 +235,10 @@ def _fetch_file(url, file_name, print_destination=True):
     initial_size = 0
     # Checking file size and displaying it alongside the download url
     n_try = 3
+    contxt = ssl._create_unverified_context()
     for ii in range(n_try):
         try:
-            data = urllib.request.urlopen(url, timeout=15.)
+            data = urllib.request.urlopen(url, timeout=15., context=contxt)
```
Comment (djhoese) [inline vispy/util/fetching.py:241]: Sorry, but please remove this. Someone else has tried this in the past. This is not the right solution and completely negates the purpose of an HTTPS URL. I should go find that old conversation about this.

**Comment by djhoese** [comment]
I just merged a PR into master that is causing conflicts with your PR, sorry. Also, that merge has successfully passed on Azure so there must be something that is different in your PR specifically.

**Comment by djhoese** [comment]
Ok I figured out why this only fails in your PR. The test that is failing is:

```

@requires_img_lib()
def test_read_write_image():
    """Test reading and writing of images"""
    fname = op.join(temp_dir, 'out.png')
    im1 = load_crate()
    imsave(fname, im1, format='png')
    with warnings.catch_warnings(record=True):  # PIL unclosed file
        im2 = imread(fname)
    assert_allclose(im1, im2)

```

The decorator at the top `requires_img_lib` is checking for either imageio or pillow. That has always been false in the past but now that you've added pillow to the installed dependencies it is running the test instead of skipping it. I'm not sure what the right solution is here. I'll have to think about it.

File: vispy/util/fetching.py (hunk 3 above)
Comment (Kartik-byte) [reply vispy/util/fetching.py:241]: I am sorry. I was just trying to see if the SSL check was the only thing causing this.
I found the issue [https://stackoverflow.com/questions/42098126/mac-osx-python-ssl-sslerror-ssl-certificate-verify-failed-certificate-verify] 
I think mac OS doesn't come preinstalled with certificates.
This post gives a way to install it but I am not sure how to do that on azure

**Comment by Kartik-byte** [comment]
Thanks soo much.
I was sooo confused on how i messed up this SSL check

**Comment by djhoese** [comment]
Besides merging with master to fix the merge conflicts, try making the following changes to `test_image`:

```
Index: vispy/io/tests/test_image.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- vispy/io/tests/test_image.py	(revision 88cc318dfee23fad947cb0f0bcee843d4e0e2a57)
+++ vispy/io/tests/test_image.py	(date 1602107174959)
@@ -5,6 +5,8 @@
 from numpy.testing import assert_array_equal, assert_allclose
 from os import path as op
 import warnings
+import sys
+import pytest
 
 from vispy.io import load_crate, imsave, imread, read_png, write_png
 from vispy.testing import requires_img_lib, run_tests_if_main
@@ -35,6 +37,8 @@
 
 
 @requires_img_lib()
+@pytest.mark.xfail(sys.version_info < (3, 7) and 'darwin' in sys.platform,
+                   reason="Python 3.6 on OSX sometimes fails to download from HTTPS.")
 def test_read_write_image():
     """Test reading and writing of images"""
     fname = op.join(temp_dir, 'out.png')

```

(the `+` indicates lines that were added, all other lines are for context/reference)

**Comment by djhoese** [comment]
Let me know if you need help with the merge conflicts (I get an email every time you push a commit :wink: )

**Comment by Kartik-byte** [comment]
I am soo sorry ,I am new to git and github and this merging conflicts has got me soo confused.
I do need help.
I am sorry😞

**Comment by djhoese** [comment]
Don't worry about it. Git can be really confusing. I just mentioned the emails to explain how I could tell you were having trouble. First, how do you have your remotes set up?

```
git remote -v
```

You'll want to have two remotes, one for the vispy main repository and one for your fork. If you cloned from your fork then you may need to do:

```
git remote add upstream git@github.com:vispy/vispy.git
git fetch upstream
```

Then do:

```
git pull upstream master
```

(I forget if you need a slash between upstream and master with that one)

Then git will complain about merge conflicts and then you follow its instructions to fix them and commit the changes and then push here. If you need any help with that just let me know.

**Comment by Kartik-byte** [comment]
Thank you soo much.
Will let you know if I face any issues.😁

**Comment by Kartik-byte** [comment]
I think I managed to fix the merge conflicts but the test is still not passing

**Comment by Kartik-byte** [comment]
I think right now vispy is using python 3.7x so checking for less than 3.7 is not working. This issue is present in newer version of python too soo should changing the condition to > 3.6 be a good idea?

**Comment by djhoese** [comment]
Ah I saw the failure for python 3.6 and assumed it was only for 3.6. If we are skipping for 3.6+ then we might as well just remove the python version part of the check since that's all vispy supports right now. I'll check with the people who build these docker build images and see if there is a solution they can implement. Also, you can ignore the appveyor failures. That is a recent development that isn't your fault.

**Comment by Kartik-byte** [comment]
That's so true. I realized that now.
Apart from appveyor, 1 travis test is also failing, while all others are passing. Both the errors looks similar to me.
I think a good fix for macos issue will be installing the required certificates in the testing machine. ref link:- [Here](https://stackoverflow.com/a/42334357)
But I don't have any idea how to do that on the test machine.
I can try if you can guide me.
Anyways ,
Thank you for guiding me through this PR . Got to learn so many things.

**Comment by djhoese** [comment]
FYI https://github.com/joerick/cibuildwheel/issues/443

**Comment by Kartik-byte** [comment]
Okay so I removed the code that skipped the test for macOS . Now with the test version of cibuildwheel you added, macOS is passing but macos_10_14  is not. Its having the same SSL error. Any idea why or how can we patch the test for macos_10_14 too?
Thank you.

**Comment by Kartik-byte** [comment]
Okay I figured it out. 
The test dev of cibuildwheel did fix that issue. Yay😄
Thanks again for helping me through this PR.
Please let me know if there is something else I can do.

**Comment by djhoese** [comment]
Oh sorry. I didn't actually mean to push those to your branch. So sorry. I misconfigured my local branch and pushed them here. I was testing this stuff somewhere else. I guess I'll fix your branch when all the cibuildwheel stuff is figured out. Sorry again and thanks for the work on this and your patience.

**Comment by Kartik-byte** [comment]
Oh okay, just noticed the new cibuildwheel test branch on vispy upstream.Thanks for the help on this.
Just a note, you might wanna remove the test skip we added for macOS in test_image.py to actually test new cibuildwheel.
Thanks again and happy to contribute.😁

**Comment by Kartik-byte** [comment]
With the new release of cibuildwheel the ssl issue has been fixed. Is it good to go now?
Please let me know if something else has to be done.
Thank You.

**Comment by Kartik-byte** [comment]
@djhoese is it good now?

**Review by djhoese — APPROVED** [review]

**Comment by djhoese** [comment]
Yes, this looks good. Thanks again.

**Pull request merged**